wireguard-go-rs/src/lib.rs

```diff
@@ -1,70 +1,265 @@
-    /// for the tunnel device and logging.
-    ///
-    /// Positive return values are tunnel handles for this specific wireguard tunnel instance.
-    /// Negative return values signify errors. All error codes are opaque.
-    #[cfg(not(target_os = "android"))]
-    pub fn wgTurnOn(
-        mtu: isize,
-        settings: *const i8,
-        fd: Fd,
-        logging_callback: Option<LoggingCallback>,
-        logging_context: *mut c_void,
-    ) -> i32;
+/// A wireguard-go tunnel
+pub struct Tunnel {
+    /// wireguard-go handle to the tunnel.
+    handle: i32,
+}
 
-    // Android
-    #[cfg(target_os = "android")]
-    pub fn wgTurnOn(
-        settings: *const i8,
-        fd: Fd,
+// NOTE: Must be kept in sync with libwg.go
+// NOTE: must be kept in sync with `result_from_code`
+// INVARIANT: Will aways be represented as a negative i32
+#[repr(i32)]
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Something went wrong.")]
+    GeneralFailure = -1,
+
+    #[error("Something went wrong, but trying again might help.")]
+    IntermittentFailure = -2,
+
+    #[error("An argument you provided was invalid.")]
+    InvalidArgument = -3,
+
+    #[error("The tunnel handle did not refer to an existing tunnel.")]
+    UnknownTunnel = -4,
+
+    #[error("The provided public key did not refer to an existing peer.")]
+    UnknownPeer = -5,
+
+    #[error("Something went wrong when enabling DAITA.")]
+    EnableDaita = -6,
+
+    #[error("`libwg` provided an unknown error code. This is a bug.")]
+    Other = i32::MIN,
+}
+
+impl Tunnel {
+    // TODO: this function is supposed to be a safe wrapper, but as clippy points out, 
+    // the logging_context is a *mut, which may unsafely be dereferenced by the callback.
+    // I'd prefer NOT to mark this functon as unsafe though...
+    pub fn turn_on(
+        #[cfg(not(target_os = "android"))]
+        mtu: isize,
+        settings: &CStr,
+        device: Fd,
         logging_callback: Option<LoggingCallback>,
         logging_context: *mut c_void,
-    ) -> i32;
+    ) -> Result<Self, Error> {
+        // SAFETY: pointer is valid for the the lifetime of this function
+        let code = unsafe {
+            ffi::wgTurnOn(
+                #[cfg(not(target_os = "android"))]
+                mtu,
 
-    // Pass a handle that was created by wgTurnOn to stop a wireguard tunnel.
-    pub fn wgTurnOff(handle: i32) -> i32;
+                settings.as_ptr(),
+                device,
+                logging_callback,
+                logging_context,
+            )
+        };
 
-    // Returns the file descriptor of the tunnel IPv4 socket.
-    pub fn wgGetConfig(handle: i32) -> *mut c_char;
+        result_from_code(code)?;
+        Ok(Tunnel { handle: code })
+    }
 
-    // Sets the config of the WireGuard interface.
-    pub fn wgSetConfig(handle: i32, settings: *const i8) -> i32;
+    pub fn turn_off(self) -> Result<(), Error> {
```

Comment:
It's a bit akward call `ffi::wgTurnOff` both on drop and also provide a manual version. Slightly outside the scope of this PR, but could the `stop` method on the `Tunnel` trait not just be removed in favour of just cleaning up on drop for all `Tunnel` implementations?

Reply:
The benefit of having a stop function is that we can explicitly handle the error in case wgTurnOff should fail

Reply:
The only place we call the stop function is in  `talpid_wireguard::WireguardMonitor::stop_tunnel`, and there we just print the error and then throw it away

Reply:
True. Could be worth refactoring at some point

Reply:
Alright, I suggest we add a TODO note and leave it for now